### PR TITLE
Added Autoscroll on new messages - FIX issue 1278

### LINKED ISSUE #1278: Messages don't scroll to the bottom
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/caprine/issues/1278)

Most of the time, when new messages are received or you switch conversations, the scroll position does not move to the bottom and you have to manually scroll to see conversations.

This is such a frustrating bug, it must only happen in certain scenarios that don't happen to the authors, because the app is so well thought through otherwise. I am not sure what is unusual about what I am doing though.

---

Caprine 2.44.0
Electron 8.0.0
darwin 19.3.0
Locale: en-US

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $50.00)

- [wilg wilg](https://issuehunt.io/u/wilg) ($50.00)

### Submitted pull Requests
- [#1295 Autoscroll (#1278) ](https://issuehunt.io/r/sindresorhus/caprine/pull/1295)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/caprine/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by fbianconi**
Hi there, I've been working in the autoscroll issue (fixes #1278) that I saw in IssueHunt. Let me know if this makes the cut, or if you need anything else.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#1278: Messages don't scroll to the bottom](https://issuehunt.io/repos/42574339/issues/1278)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

**Comment by sindresorhus** [comment]
This should not be an option. This fixes a bug.

**Comment by sindresorhus** [comment]
And make the PR title more descriptive.

File: source/browser.ts [hunk 1]
```diff
@@ -540,6 +540,41 @@ function insertionListener(event: AnimationEvent): void {
 	}
 }
 
+async function observeAutoscroll(): Promise<void> {
+	const mainElement = await elementReady<HTMLElement>('body>div>div>div', {stopOnDomReady: false});
```
Comment (sindresorhus) [inline source/browser.ts]: The selector here is too fragile.

File: source/browser.ts [hunk 2]
```diff
@@ -540,6 +540,41 @@ function insertionListener(event: AnimationEvent): void {
...
+	if (mainElement) {
```
Comment (sindresorhus) [inline source/browser.ts]: Do an early return instead.

File: source/browser.ts [hunk 3]
```diff
@@ -540,6 +540,41 @@ function insertionListener(event: AnimationEvent): void {
...
+		const hookMessageObserver = async (): Promise<void> => {
+			const chatElement = await elementReady<HTMLElement>(
+				'[role=presentation] .scrollable > div > div > div > div:nth-child(3)', {stopOnDomReady: false}
```
Comment (sindresorhus) [inline source/browser.ts]: The selector here is too fragile. Messenger could change a tiny detail and this could break.

File: source/browser.ts [hunk 4]
```diff
@@ -540,6 +540,41 @@ function insertionListener(event: AnimationEvent): void {
...
+			);
+			if (chatElement) {
+				const messageObserver = new MutationObserver((records: MutationRecord[]) => {
+					const newMessages: MutationRecord[] = records.filter(r => {
```
Comment (sindresorhus) [inline source/browser.ts]:
```suggestion
					const newMessages: MutationRecord[] = records.filter(record => {
```

File: source/browser.ts [hunk 5]
```diff
@@ -540,6 +540,41 @@ function insertionListener(event: AnimationEvent): void {
...
+						// The mutation it's an addition
```
Comment (sindresorhus) [inline source/browser.ts]:
```suggestion
						// The mutation is an addition
```

File: source/browser.ts [hunk 6]
```diff
@@ -579,6 +614,9 @@ document.addEventListener('DOMContentLoaded', async () => {
 
 	// Disable autoplay if set in settings
 	toggleVideoAutoplay();
+
+	// Hook auto scroll observer
```
Comment (sindresorhus) [inline source/browser.ts]:
```suggestion
	// Hook auto-scroll observer
```

**Comment by fbianconi** [comment]
I've made the changes requested. I hope this helps. Let me know if there are still any problems.

**Comment by sindresorhus** [comment]
In a conversation, if I scroll far up, switch to a different conversation, and then back again, it does not scroll down to the most recent message.

File: source/browser.ts [hunk 7]
```diff
@@ -540,6 +540,43 @@ function insertionListener(event: AnimationEvent): void {
...
+	if (!mainElement) {
+		return;
+	}
+
+	const hookMessageObserver = async (): Promise<void> => {
+		const chatElement = await elementReady<HTMLElement>(
+			'[role=presentation] .scrollable [role = region] > div[id ^= "js_"]', {stopOnDomReady: false}
+		);
+		if (chatElement) {
+			const messageObserver = new MutationObserver((r: MutationRecord[]) => {
```
Comment (sindresorhus) [inline source/browser.ts:555]:
```suggestion
			const messageObserver = new MutationObserver((record: MutationRecord[]) => {
```

File: source/browser.ts [hunk 8]
```diff
@@ -540,6 +540,43 @@ function insertionListener(event: AnimationEvent): void {
...
+				const newMessages: MutationRecord[] = r.filter(r => {
+					// The mutation is an addition
+					return r.addedNodes.length > 0 &&
```
Comment (sindresorhus) [inline source/browser.ts:558]: Don't use one character variable names.

**Comment by alekksander** [comment]
this is quite annoying when browsing previous messages in search for something. it also doesn't work once coming back from gallery after viewing shared pictures – after closing picture conversation view is somewhere „in the middle”.

**Comment by sindresorhus** [comment]
Fixed by https://github.com/sindresorhus/caprine/pull/1382

**Pull request closed without merging**